bundles/io/org.eclipse.smarthome.io.json/src/main/java/org/eclipse/smarthome/io/json/gson/GsonBindingService.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.io.json.gson;
+
+import java.io.Reader;
+import java.io.Writer;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.io.json.JsonBindingService;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * Gson implementation of a JsonBindingLayer.
+ *
+ * @author Flavio Costa - Initial contribution.
+ *
+ * @param <T> Type to be serialized or deserialized.
+ */
+public class GsonBindingService<T> implements JsonBindingService<T> {
+
+    /**
+     * Builder for Gson.
+     */
+    private GsonBuilder builder;
+
+    /**
+     * Gson instance used for (de)serialization.
+     */
+    private Gson gson;
+
+    @Override
+    public void toJson(T object, Type genericType, Writer writer) {
+        getGson().toJson(object, genericType, writer);
+    }
+
+    @Override
+    public T fromJson(Reader reader, Type genericType) {
+        return getGson().fromJson(reader, genericType);
+    }
+
+    @Override
+    public void setFormattedOutput(boolean prettyPrinting) {
+        if (prettyPrinting) {
+            builder = getGsonBuilder().setPrettyPrinting();
+        }
+    }
+
+    /**
+     * Registers a set of type adapters into the GsonBuilder.
+     *
+     * @param entries Entries to be registered.
+     */
+    protected void registerTypeAdapters(Set<Map.Entry<Class<?>, Object>> entries) {
+        entries.forEach(e -> getGsonBuilder().registerTypeAdapter(e.getKey(), e.getValue()));
+    }
+
+    /**
+     * Checks if the Gson instance has been created, then return a new instance or reuse the existing one.
+     *
+     * @return Created or previously existing Gson instance.
+     */
+    protected Gson getGson() {
```

Comment:
Having this only as `protected` will make the internal implementation, here `Gson`, leak out. Please make this `private`.

Reply:
This is protected to allow subclasses to overwrite this behavior if needed. This method is on the `GsonBindingService` class only, which is already Gson-specific. It does not appear in the interface.